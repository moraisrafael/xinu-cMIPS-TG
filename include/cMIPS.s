
	# see vhdl/packageMemory.vhd for addresses
        .set x_INST_BASE_ADDR,0x00000000
        .set x_INST_MEM_SZ,0x00020000

        .set x_DATA_BASE_ADDR,0x00080000
        .set x_DATA_MEM_SZ,0x00080000
	
        .set x_IO_BASE_ADDR,0x3c000000
        .set x_IO_MEM_SZ,0x00002000
	.set x_IO_ADDR_RANGE,0x00000020

	.set x_SDRAM_BASE_ADDR,0x04000000
	.set x_SDRAM_MEM_SZ,0x02000000
	
	.set HW_counter_addr,(x_IO_BASE_ADDR +  5 * x_IO_ADDR_RANGE)
	.set HW_FPU_addr,    (x_IO_BASE_ADDR +  6 * x_IO_ADDR_RANGE)
	.set HW_uart_addr,   (x_IO_BASE_ADDR +  7 * x_IO_ADDR_RANGE)
	.set HW_dsp7seg_addr,(x_IO_BASE_ADDR +  9 * x_IO_ADDR_RANGE)
	.set HW_keybd_addr,  (x_IO_BASE_ADDR + 10 * x_IO_ADDR_RANGE)
	.set HW_lcd_addr,    (x_IO_BASE_ADDR + 11 * x_IO_ADDR_RANGE)
	.set HW_sdc_addr,    (x_IO_BASE_ADDR + 12 * x_IO_ADDR_RANGE)

	# see vhdl/packageMemory.vhd for addresses
	.set x_EXCEPTION_0000,0x00000130
	.set x_EXCEPTION_0100,0x00000200
	.set x_EXCEPTION_0180,0x00000280
	.set x_EXCEPTION_0200,0x00000400
	.set x_EXCEPTION_BFC0,0x00000680
	.set x_ENTRY_POINT,   0x00000700


	.set c0_index,   $0
	.set c0_random,  $1
	.set c0_entrylo0,$2
	.set c0_entrylo1,$3
	.set c0_context ,$4
	.set c0_pagemask,$5
	.set c0_wired,   $6
	.set c0_badvaddr,$8
	.set c0_count   ,$9
	.set c0_entryhi ,$10
	.set c0_compare ,$11
	.set c0_status  ,$12
	.set c0_cause   ,$13
	.set c0_epc,     $14
	.set c0_config,  $16
	.set c0_config_f0,0
	.set c0_config_f1,1
	.set c0_lladdr,  $17
	.set c0_errorpc, $30

	
	# reset: COP0 present, at exception level, all else disabled
	.set c0_status_reset,0x10000002
	
	# normal state: COP0 present, user mode, all IRQs enabled
	.set c0_status_normal,0x1000ff11

	# reset: COUNTER stopped, use special interrVector, no interrupts
	.set c0_cause_reset, 0x0880007c


	# not used in XINU, RH
#	.equ led_RED, 0x4000
#	.equ led_GRE, 0x2000
#	.equ led_BLU, 0x1000
#	.equ led_OFF, 0x0FFF

