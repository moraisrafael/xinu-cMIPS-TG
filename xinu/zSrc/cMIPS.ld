OUTPUT_ARCH(mips)
ENTRY(_start)

MEMORY
{
        rom (rx)  : ORIGIN = 0x, /* x_INST_BASE_ADDR */
                    LENGTH = 0x, /* x_INST_MEM_SZ */
        ram (!rx) : ORIGIN = 0x, /* x_DATA_BASE_ADDR */
                    LENGTH = 0x, /* x_DATA_MEM_SZ */
        io (!rx)  : ORIGIN = 0x0f000000, /* not used, do not remove       */
                    LENGTH = 0x00020000  /* edMemory.sh needs thess lines */
}

SECTIONS
{
       .text :
       {
          *(.text .text.*)
          _etext = . ;         /* end of text constant (from Xinu) */
       }                                   > rom

       .data : 
       {
          _bdata = . ;         /* begin of data constant */
          *(.data .data.*) 
          _edata = . ;         /* end of data constant  (from Xinu) */
       }                                   > ram
       .data1   : { *(.data1) }            > ram
       .rodata  : { *(.rodata .rodata.*) } > ram
       .rodata1 : { *(.rodata1) }          > ram
       .lit8    : { *(.lit8) }             > ram
       .lit4    : { *(.lit4) }             > ram
       .sdata   : { *(.sdata .sdata.*) }   > ram
       .sbss    : { *(.sbss .sbss.*) *(.scommon .scommon.*) }  > ram
       .bss :
       {
          *(.bss .bss.*) *(COMMON) 
          _end = . ;           /* end of image constant  (from Xinu) */
       }                                   > ram

       end_RAM = 0x; /* x_DATA_MEM_SZ */
       half_RAM = (end_RAM / 2);
       base_TP = ( _bdata + half_RAM );
       .TP base_TP : { *(.TP) }                  > ram
}
