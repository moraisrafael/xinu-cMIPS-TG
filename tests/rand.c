/***************************************************************
 * check randomness of rand() [after Kelley & Phol pp 81]
 ***************************************************************/

#ifdef cMIPS
  #include "cMIPS.h"
#else
  #include <stdio.h>
  #include <stdlib.h>
#endif

#define ABS(x) ( ((x) < 0) ? -(x) : (x) )

 int Random( int *m_z,  int *m_w) { // from wikipedia
  *m_z = 36969 * (*m_z & 65535) + (*m_z >> 16);
  *m_w = 18000 * (*m_w & 65535) + (*m_w >> 16);
  return ((*m_z << 16) + *m_w);  /* 32-bit result */
}

#define MAX_NUM 100
#define SHOW (MAX_NUM / 5)

void main(void){
  int i, above, below, maxdif, newv;
  int median;
  
  int m_w = 177;    /* must not be zero, nor 0x464fffff */
  int m_z = 311;    /* must not be zero, nor 0x9068ffff */

  above = below = maxdif = 0;

  median = 0xffffffff;  // median ===> (2^32 - 1)

#ifdef cMIPS
  // print(median); to_stdout('\n');
#else
  // printf("%08x\n\n", median);
#endif

  for( i=1; i<= MAX_NUM; i++ ){
    if( Random(&m_z, &m_w) <= median ) below++ ;
    else above++ ;
    if( (newv = ABS(above - below)) > maxdif ) maxdif = newv;
    if( (i % SHOW) == 0 ) {
#ifdef cMIPS
      print(i); print(above); print(below); print(maxdif); to_stdout('\n');
#else
      printf("%08x\n%08x\n%08x\n%08x\n\n", i, above, below, maxdif);
#endif
    }
  }
}

