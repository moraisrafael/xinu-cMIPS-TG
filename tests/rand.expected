00000014
00000008
0000000c
00000005

00000028
00000014
00000014
00000005

0000003c
00000022
0000001a
00000008

00000050
0000002d
00000023
0000000c

00000064
00000038
0000002c
0000000d

